torchgen/model.py

```diff
@@ -1201,7 +1201,15 @@ def __post_init__(self) -> None:
                 ), "out= ops that accept tensor lists as out arguments "
                 "are expected to have no return type (since you can't do method chaining on them)"
             else:
-                assert len(self.arguments.out) == len(
+                # mutable keyward arguments whose name has _scratch_ prefix are
+                # scratch tensors for memory planning and should not be returned
+                assert len(
+                    [
+                        arg
+                        for arg in self.arguments.out
+                        if not arg.name.startswith("_scratch_")
```

Comment:
can we check the schema kind here? I'd rather bake in the invariant of `"_scratch_" -> SchemaKind.scratch` once, and just rely on the schema kind everywhere else.

Reply:
Hmm, but the code here does not only check if an op is an out variant or a scratch op. Be more specific,
1. if the op s a out variant op then `len(self.arguments.out)` should equal to `len(self.returns)`
2. if the op is a scratch op, then we should skip the _scratch_ tensor arguments when comparing the number of out arguments against the number of return items.

Case 1 is just handled naturally since it's a special case of 2.

Even if we check for SchemaKind here, we still need to filter out _scratch_ tensors in case 2 by checking the argument name.

Let me know if I misunderstood.

Reply:
So the reason here is _scratch_ tensors are not returned.

Reply:
oh yep I misunderstood. Sounds good!

We might wanna make the `_scratch_` check more first-class, and introduce a method like `Argument.is_scratch()` (just so the string `_scratch_` isn't baked into the codegen in multiple places).